hadoop-tools/hadoop-msgraph/src/main/java/org/apache/hadoop/security/msgraph/oauth2/AzureADAuthenticator.java

```diff
@@ -0,0 +1,246 @@
+      conn.setConnectTimeout((int) TimeUnit.SECONDS.toMillis(30));
+      String requestId;
+
+      conn.setRequestProperty(HttpHeader.CONNECTION.asString(), "close");
+      conn.setDoOutput(true);
+      conn.getOutputStream().write(payload.getBytes(StandardCharsets.UTF_8));
+
+      int httpResponseCode = conn.getResponseCode();
+      requestId = conn.getHeaderField("x-ms-request-id");
+      String responseContentType = conn.getHeaderField(
+          HttpHeader.CONTENT_TYPE.asString());
+      long responseContentLength = conn.getHeaderFieldLong(
+          HttpHeader.CONTENT_LENGTH.asString(), 0L);
+      requestId = requestId == null ? "" : requestId;
+
+      if (httpResponseCode != HttpStatus.OK_200 ||
+          !responseContentType.startsWith(
+              MimeTypes.Type.APPLICATION_JSON.asString()) ||
+          responseContentLength <= 0L) {
+        String responseBody = consumeInputStream(conn.getInputStream());
+        String proxies = "none";
+        String httpProxy = System.getProperty("http.proxy");
+        String httpsProxy = System.getProperty("https.proxy");
+        if (httpProxy != null || httpsProxy != null) {
+          proxies = "http:" + httpProxy + ";https:" + httpsProxy;
+        }
+
+        String logMessage = String.format("AADToken: HTTP connection failed " +
+                "for getting token from AzureAD. Http response: %d %s " +
+                ", Content-Type: %s, Content-Length: %d, Request ID: %s" +
+                ", Proxies: %s, First 1K of Body: %s.",
+            httpResponseCode, conn.getResponseMessage(), responseContentType,
+            responseContentLength, requestId, proxies, responseBody);
+        LOG.debug(logMessage);
+        throw new HttpException(httpResponseCode, requestId, logMessage);
+      }
+
+      InputStream httpResponseStream = conn.getInputStream();
+      token = parseTokenFromStream(httpResponseStream);
+    } finally {
+      if (conn != null) {
+        conn.disconnect();
+      }
+    }
+
+    return token;
+  }
+
+  private static String consumeInputStream(InputStream inStream)
+      throws IOException {
+    final int bufferSize = 1024;
+
+    int totalBytesRead = 0;
+    int bytesRead;
+    ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+
+    do {
+      byte[] b = new byte[bufferSize];
+      bytesRead = inStream.read(b, totalBytesRead, bufferSize);
+      if (bytesRead > 0) {
+        totalBytesRead += bytesRead;
+        bytes.write(b);
+      }
+    } while (bytesRead > 0);
+
+    return new String(bytes.toByteArray(), 0, totalBytesRead);
+  }
+
+  /**
+   * Parse a token from an HTTP stream.
+   * @param httpResponseStream HTTP response stream.
+   */
+  private static AzureADToken parseTokenFromStream(
+      InputStream httpResponseStream) throws IOException {
+    AzureADToken token;
+
+    try {
+      Integer expiryPeriod = null;
+      String accessToken = null;
+
+      JsonFactory jf = new JsonFactory();
```

Comment:
check the content type for being JSON, I have had to feel to many support calls related to ADLS and Proxies and its naive parser.